modules/add-contextual-data/add-contextual-data-glob-selector.c

```diff
@@ -0,0 +1,176 @@
+  g_pattern_spec_free(gs->expr);
+}
+
+typedef struct _AddContextualDataGlobSelector
+{
+  AddContextualDataSelector super;
+  GArray *globs;
+  LogTemplate *glob_template;
+} AddContextualDataGlobSelector;
+
+static GArray *
+_create_globs_array(void)
+{
+  GArray *array = g_array_new(FALSE, TRUE, sizeof(GlobExpression));
+  g_array_set_clear_func(array, (void (*)(void *)) glob_expression_clear);
+  return array;
+}
+
+static void
+_populate_globs_array(AddContextualDataGlobSelector *self, GList *ordered_selectors)
+{
+  for (GList *l = ordered_selectors; l; l = l->next)
+    {
+      const gchar *selector = (const gchar *) l->data;
+      GlobExpression gs;
+
+      glob_expression_populate(&gs, selector);
+      g_array_append_val(self->globs, gs);
+    }
+}
+
+static GArray *
+_clone_globs_array(GArray *src)
+{
+  GArray *dst = _create_globs_array();
+
+  for (gint i = 0; i < src->len; i++)
+    {
+      GlobExpression *src_expr = &g_array_index(src, GlobExpression, i);
+      GlobExpression dst_expr;
+
+      glob_expression_populate(&dst_expr, src_expr->pattern);
+      g_array_append_val(dst, dst_expr);
+    }
+  return dst;
+}
+
+static const gchar *
+_find_first_matching_glob(AddContextualDataGlobSelector *self, LogMessage *msg)
+{
+  GString *string = scratch_buffers_alloc();
+  GString *string_reversed = scratch_buffers_alloc();
+
+  log_template_format(self->glob_template, msg, NULL, LTZ_LOCAL, 0, NULL, string);
+  g_string_assign(string_reversed, string->str);
+  g_strreverse(string_reversed->str);
+  for (gint i = 0; i < self->globs->len; i++)
+    {
+      GlobExpression *gs = &g_array_index(self->globs, GlobExpression, i);
+      gboolean result = g_pattern_match(gs->expr, string->len, string->str, string_reversed->str);
+
+      msg_trace("add-contextual-data(): Evaluating glob against message",
+                evt_tag_str("glob-template", self->glob_template->template),
+                evt_tag_str("string", string->str),
+                evt_tag_str("pattern", gs->pattern),
+                evt_tag_int("matched", result));
+      if (result)
+        return gs->pattern;
+    }
+
+  return NULL;
+}
+
+static gboolean
+_init(AddContextualDataSelector *s, GList *ordered_selectors)
+{
+  AddContextualDataGlobSelector *self = (AddContextualDataGlobSelector *)s;
+
+  _populate_globs_array(self, ordered_selectors);
+
+  return TRUE;
+}
+
+static gchar *
+_resolve(AddContextualDataSelector *s, LogMessage *msg)
+{
+  AddContextualDataGlobSelector *self = (AddContextualDataGlobSelector *)s;
+
+  return g_strdup(_find_first_matching_glob(self, msg));
+}
+
+static void
+_free(AddContextualDataSelector *s)
+{
+  AddContextualDataGlobSelector *self = (AddContextualDataGlobSelector *)s;
+
+  g_array_free(self->globs, TRUE);
+}
+
+static AddContextualDataSelector *_clone(AddContextualDataSelector *s,
+    GlobalConfig *cfg);
+
+static AddContextualDataGlobSelector *
+_create_empty_add_contextual_data_glob_selector(LogTemplate *glob_template)
```

Comment:
A similar function was refactored to an `..._init_instance()` type function in a previous commit. Please follow that methodology here, too.

Reply:
oops, I've fixed that up in the existing code because I didn't like it when modeling my own code. And then didn't follow that up in my own.

embarassing.

now it's fixed.